File: client-android-kotlin/messaging/src/main/java/com/voximplant/demos/messaging/repository/utils/CustomData.kt
package com.voximplant.demos.messaging.repository.utils

import com.voximplant.demos.messaging.utils.permissions.Permissions

typealias CustomData = MutableMap<String, Any?>

var CustomData.type: String?
    get() = this[TYPE] as String?
    set(value) = this.set(TYPE, value)

var CustomData.image: String?
    get() = this[IMAGE] as String?
    set(value) = this.set(IMAGE, value)

var CustomData.chatDescription: String?
    get() = this[DESCRIPTION] as String?
    set(value) = this.set(DESCRIPTION, value)

var CustomData.status: String?
    get() = this[STATUS] as String?
    set(value) = this.set(STATUS, value)

var CustomData.permissions: Permissions?
    @Suppress("UNCHECKED_CAST")
    get() = this[PERMISSIONS] as Permissions?
    set(value) = this.set(PERMISSIONS, value)

private const val TYPE = "type"
private const val IMAGE = "image"
private const val DESCRIPTION = "description"
private const val STATUS = "status"
private const val PERMISSIONS = "permissions"

File: client-ios-swift/Messaging/Data/Local/Entity/CustomData.swift
/*
*  Copyright (c) 2011-2019, Zingaya, Inc. All rights reserved.
*/

import Foundation

typealias CustomData = [String: NSObject]

extension CustomData {
    var type: Conversation.ConversationType {
        get {
            switch self["type"] as? String {
            case "direct":
                return .direct
            case "chat":
                return .chat
            case "channel":
                return .channel
            default:
                return .chat
            }
        }
        set {
            var typeString: NSString?
            switch newValue {
            case .direct:
                typeString = "direct"
            case .chat:
                typeString = "chat"
            case .channel:
                typeString = "channel"
            }
            self["type"] = typeString
        }
    }
    
    var image: String? {
        get { self["image"] as? String }
        set { self["image"] = newValue as NSString? }
    }
    
    var chatDescription: String? {
        get { self["description"] as? String }
        set { self["description"] = newValue as NSString? }
    }
    
    var status: String? {
        get { self["status"] as? String }
        set { self["status"] = newValue as NSString? }
    }
    
    var permissions: NSDictionary? {
        get { self["permissions"] as? NSDictionary }
        set { self["permissions"] = newValue }
    }
}
